internal/services/eventgrid/eventgrid_partner_namespace_resource.go

```diff
@@ -0,0 +1,331 @@
+	return map[string]*pluginsdk.Schema{
+		"endpoint": {
+			Type:     pluginsdk.TypeString,
+			Computed: true,
+		},
+	}
+}
+
+func (r EventGridPartnerNamespaceResource) ModelObject() interface{} {
+	return &EventGridPartnerNamespaceResourceModel{}
+}
+
+func (EventGridPartnerNamespaceResource) ResourceType() string {
+	return "azurerm_eventgrid_partner_namespace"
+}
+
+func (r EventGridPartnerNamespaceResource) Create() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.EventGrid.PartnerNamespaces
+
+			subscriptionId := metadata.Client.Account.SubscriptionId
+
+			var config EventGridPartnerNamespaceResourceModel
+			if err := metadata.Decode(&config); err != nil {
+				return fmt.Errorf("decoding: %+v", err)
+			}
+			id := partnernamespaces.NewPartnerNamespaceID(subscriptionId, config.ResourceGroup, config.PartnerNamespaceName)
+
+			existing, err := client.Get(ctx, id)
+			if err != nil && !response.WasNotFound(existing.HttpResponse) {
+				return fmt.Errorf("checking for presence of existing %s: %+v", id, err)
+			}
+			if !response.WasNotFound(existing.HttpResponse) {
+				return metadata.ResourceRequiresImport(r.ResourceType(), id)
+			}
+
+			publicNetworkAccess := partnernamespaces.PublicNetworkAccessEnabled
+			if !config.PublicNetworkAccessEnabled {
+				publicNetworkAccess = partnernamespaces.PublicNetworkAccessDisabled
+			}
+
+			param := partnernamespaces.PartnerNamespace{
+				Location: config.Location,
+				Properties: &partnernamespaces.PartnerNamespaceProperties{
+					DisableLocalAuth:                    pointer.To(!config.LocalAuthEnabled),
+					InboundIPRules:                      expandPartnerInboundIPRules(config.InboundIPRules),
+					PartnerRegistrationFullyQualifiedId: pointer.To(config.PartnerRegistrationFullyQualifiedID),
+					PartnerTopicRoutingMode:             pointer.To(partnernamespaces.PartnerTopicRoutingMode(config.PartnerTopicRoutingMode)),
+					PublicNetworkAccess:                 pointer.To(publicNetworkAccess),
+				},
+				Tags: pointer.To(config.Tags),
+			}
+
+			if err := client.CreateOrUpdateThenPoll(ctx, id, param); err != nil {
+				return fmt.Errorf("creating %s: %+v", id, err)
+			}
+
+			metadata.SetID(id)
+			return nil
+		},
+	}
+}
+
+func expandPartnerInboundIPRules(input []PartnerInboundIpRuleModel) *[]partnernamespaces.InboundIPRule {
+	if len(input) == 0 {
+		return nil
+	}
+
+	ipRules := make([]partnernamespaces.InboundIPRule, 0)
+	for _, v := range input {
+		ipRules = append(ipRules, partnernamespaces.InboundIPRule{
+			Action: pointer.To(partnernamespaces.IPActionType(v.Action)),
+			IPMask: pointer.To(v.IpMask),
+		})
+	}
+	return &ipRules
+}
+
+func flattenPartnerInboundIPRules(ipRules *[]partnernamespaces.InboundIPRule) []PartnerInboundIpRuleModel {
+	output := make([]PartnerInboundIpRuleModel, 0)
+
+	if ipRules == nil || len(*ipRules) == 0 {
```

Comment:
Minor, but we can remove the `len()` check here as looping over a slice of len == 0 will result in the same outcome

```suggestion
	if ipRules == nil {
```